working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+  - If `C(...)` is a constructor invocation then `C` must be a shareable class.
+  - If `g` is a reference to a global variable then `g` must be `shared`.
+
+> [!NOTE]
+>
+> You can pass non-`Shareable` values to `shared` methods but you can't do
+> anything useful with them because you can't touch their state. In other words
+> non-`Shareable` types become opaque within `shared` methods.
+
+This approach seems promising on the surface, but quickly hits issues:
+
+- It's unclear how to treat `Object` members like `toString`, `operator ==` and
+  `get hashCode`. These can't be marked as `shared` but should be accessible to
+  both `shared` and non-`shared` code.
+- It's unclear how to treat function expression invocations:
+  - Function types don't encode necessary separation between `shared` and
+    non-`shared` functions.
+  - Methods like `List<T>.forEach` pose challenge because they should be usable
+    in both `shared` and non-`shared` contexts.
+
+This makes us think that language changes required to achieve sound compile time
+delineation between `shared` and isolate worlds are too complicated to be worth
+it.
+
+### Upgrading `dart:ffi`
+
+Introduction of _shared isolate_ allows to finally address the problem of native
+code invoking Dart callbacks from arbitrary threads.
+[`NativeCallable`](https://api.dart.dev/dev/3.3.0-246.0.dev/dart-ffi/NativeCallable-class.html)
+can be extended with the corresponding constructor:
+
+```dart
+class NativeCallable<T extends Function> {
+  /// Constructs a [NativeCallable] that can be invoked from any thread.
+  ///
+  /// When the native code invokes the function [nativeFunction], the corresponding
+  /// [callback] will be synchronously executed on the same thread within a
+  /// shared isolate corresponding to the current isolate group.
+  ///
+  /// [callback] must be [Shareable] that is: all variables it captures must
+  /// have shareable static type.
+  external factory NativeCallable.shared(
+    @DartRepresentationOf("T") Function callback,
+    {Object? exceptionalReturn});
+}
+```
+
+The function pointer returned by `NativeCallable.shared(...).nativeFunction`
+will be bound to an isolate group which produced it using the same trampoline
+mechanism FFI uses to create function pointers from closures. Returned function
+pointer can be called by native code from any thread. It does not require
+exclusive access to a specific isolate and thus avoids interoperability pitfalls
+associated with that:
+
+- No need to block native caller and wait for the target isolate to become
+  available.
+- Clear semantics of globals:
+  - `shared` global state is accessible and independent from the current thread;
+  - accessing non-`shared` state will throw an `IsolationError`.
+
+#### Shared Libraries created from Dart code
+
+An introduction of _shared isolate_ allows us to adjust our deployment story and
+make it simpler to create and use shared libraries from Dart code.
```

Comment:
shared library is dynamic library or static library?

Reply:
It could be either depending on the setup. It does not really matter for the purpose of this section. I will rewrite this a bit.